File: deno/packages/providers/audio/google_tts.ts
/**
 * Google Cloud Text-to-Speech API client.
 *
 * Equivalent to Rust's `brainwires_providers::google_tts` module.
 *
 * Returns base64-encoded audio content (per Google's API contract);
 * consumers can decode with `atob` + TextEncoder or Uint8Array conversion.
 */

import { RateLimiter } from "../rate_limiter.ts";

export const GOOGLE_TTS_API_BASE = "https://texttospeech.googleapis.com/v1";

/** Text input for synthesis — set exactly one of `text` or `ssml`. */
export interface GoogleTtsInput {
  text?: string;
  ssml?: string;
}

/** Voice selection parameters. */
export interface GoogleTtsVoiceSelection {
  languageCode: string;
  name?: string;
  /** "MALE" | "FEMALE" | "NEUTRAL". */
  ssmlGender?: string;
}

/** Audio configuration. */
export interface GoogleTtsAudioConfig {
  /** "LINEAR16" | "MP3" | "OGG_OPUS" | "MULAW" | "ALAW". */
  audioEncoding: string;
  /** 0.25 – 4.0. */
  speakingRate?: number;
  /** -20.0 – 20.0. */
  pitch?: number;
  sampleRateHertz?: number;
}

/** Synthesize request. */
export interface GoogleTtsSynthesizeRequest {
  input: GoogleTtsInput;
  voice: GoogleTtsVoiceSelection;
  audioConfig: GoogleTtsAudioConfig;
}

/** Synthesize response. */
export interface GoogleTtsSynthesizeResponse {
  /** Base64-encoded audio. */
  audioContent: string;
}

/** A single voice entry. */
export interface GoogleTtsVoiceEntry {
  languageCodes: string[];
  name: string;
  ssmlGender?: string;
  naturalSampleRateHertz?: number;
}

/** Voices list response. */
export interface GoogleTtsVoicesResponse {
  voices: GoogleTtsVoiceEntry[];
}

function serializeRequest(req: GoogleTtsSynthesizeRequest): Record<string, unknown> {
  const input: Record<string, unknown> = {};
  if (req.input.text !== undefined) input.text = req.input.text;
  if (req.input.ssml !== undefined) input.ssml = req.input.ssml;

  const voice: Record<string, unknown> = { languageCode: req.voice.languageCode };
  if (req.voice.name !== undefined) voice.name = req.voice.name;
  if (req.voice.ssmlGender !== undefined) voice.ssmlGender = req.voice.ssmlGender;

  const ac: Record<string, unknown> = { audioEncoding: req.audioConfig.audioEncoding };
  if (req.audioConfig.speakingRate !== undefined) ac.speakingRate = req.audioConfig.speakingRate;
  if (req.audioConfig.pitch !== undefined) ac.pitch = req.audioConfig.pitch;
  if (req.audioConfig.sampleRateHertz !== undefined) ac.sampleRateHertz = req.audioConfig.sampleRateHertz;

  return { input, voice, audioConfig: ac };
}

/** Exposed for tests. */
export const _serializeRequest = serializeRequest;

/** Google Cloud TTS API client. */
export class GoogleTtsClient {
  readonly base_url: string;
  private readonly api_key: string;
  private rate_limiter: RateLimiter | null = null;

  constructor(api_key: string, base_url: string = GOOGLE_TTS_API_BASE) {
    this.api_key = api_key;
    this.base_url = base_url;
  }

  withRateLimit(requests_per_minute: number): this {
    this.rate_limiter = new RateLimiter(requests_per_minute);
    return this;
  }

  private async acquire(): Promise<void> {
    if (this.rate_limiter) await this.rate_limiter.acquire();
  }

  /** Synthesize speech from text. Returns base64-encoded audio content. */
  async synthesize(req: GoogleTtsSynthesizeRequest): Promise<GoogleTtsSynthesizeResponse> {
    await this.acquire();
    const res = await fetch(`${this.base_url}/text:synthesize`, {
      method: "POST",
      headers: {
        "X-Goog-Api-Key": this.api_key,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(serializeRequest(req)),
    });
    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new Error(`Google TTS API error (${res.status}): ${body}`);
    }
    return await res.json() as GoogleTtsSynthesizeResponse;
  }

  /** List available voices. */
  async listVoices(language_code?: string): Promise<GoogleTtsVoicesResponse> {
    await this.acquire();
    const url = language_code
      ? `${this.base_url}/voices?languageCode=${encodeURIComponent(language_code)}`
      : `${this.base_url}/voices`;
    const res = await fetch(url, {
      method: "GET",
      headers: { "X-Goog-Api-Key": this.api_key },
    });
    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new Error(`Google TTS voices API error (${res.status}): ${body}`);
    }
    return await res.json() as GoogleTtsVoicesResponse;
  }
}

File: crates/brainwires-hardware/src/audio/api/google_tts.rs
use std::sync::Arc;

use async_trait::async_trait;
use base64::Engine as _;
use futures::stream::BoxStream;

use brainwires_provider_speech::google_tts::{
    GoogleTtsAudioConfig, GoogleTtsClient, GoogleTtsInput, GoogleTtsSynthesizeRequest,
    GoogleTtsVoiceSelection,
};

use crate::audio::error::{AudioError, AudioResult};
use crate::audio::tts::TextToSpeech;
use crate::audio::types::{
    AudioBuffer, AudioConfig, OutputFormat, SampleFormat, TtsOptions, Voice,
};

/// Google Cloud TTS text-to-speech implementation.
///
/// Wraps a [`GoogleTtsClient`] from `brainwires-provider`.
pub struct GoogleTts {
    client: Arc<GoogleTtsClient>,
}

impl GoogleTts {
    /// Create a new Google TTS client.
    pub fn new(api_key: impl Into<String>) -> Self {
        let client = Arc::new(GoogleTtsClient::new(api_key));
        Self { client }
    }

    /// Create from an existing [`GoogleTtsClient`].
    pub fn from_client(client: Arc<GoogleTtsClient>) -> Self {
        Self { client }
    }
}

/// Map our [`OutputFormat`] to a Google Cloud audio encoding string.
fn format_to_encoding(format: OutputFormat) -> &'static str {
    match format {
        OutputFormat::Wav | OutputFormat::Pcm => "LINEAR16",
        OutputFormat::Mp3 => "MP3",
        OutputFormat::Opus => "OGG_OPUS",
        OutputFormat::Flac => "FLAC",
    }
}

#[async_trait]
impl TextToSpeech for GoogleTts {
    fn name(&self) -> &str {
        "google-tts"
    }

    async fn list_voices(&self) -> AudioResult<Vec<Voice>> {
        let resp = self
            .client
            .list_voices(None)
            .await
            .map_err(|e| AudioError::Api(format!("Google TTS list_voices: {e}")))?;

        let voices = resp
            .voices
            .into_iter()
            .map(|v| Voice {
                id: v.name.clone(),
                name: Some(v.name),
                language: v.language_codes.first().cloned(),
            })
            .collect();

        Ok(voices)
    }

    async fn synthesize(&self, text: &str, options: &TtsOptions) -> AudioResult<AudioBuffer> {
        let language = options
            .voice
            .language
            .clone()
            .unwrap_or_else(|| "en-US".to_string());

        let req = GoogleTtsSynthesizeRequest {
            input: GoogleTtsInput {
                text: Some(text.to_string()),
                ssml: None,
            },
            voice: GoogleTtsVoiceSelection {
                language_code: language,
                name: Some(options.voice.id.clone()),
                ssml_gender: None,
            },
            audio_config: GoogleTtsAudioConfig {
                audio_encoding: format_to_encoding(options.output_format).to_string(),
                speaking_rate: options.speed,
                pitch: None,
                sample_rate_hertz: Some(24000),
            },
        };

        let resp = self
            .client
            .synthesize(&req)
            .await
            .map_err(|e| AudioError::Api(format!("Google TTS synthesize: {e}")))?;

        // Google returns base64-encoded audio content.
        let decoded = base64::engine::general_purpose::STANDARD
            .decode(&resp.audio_content)
            .map_err(|e| AudioError::Format(format!("base64 decode error: {e}")))?;

        match options.output_format {
            OutputFormat::Wav => crate::audio::wav::decode_wav(&decoded),
            OutputFormat::Pcm => {
                // LINEAR16 returns raw 16-bit signed little-endian PCM.
                let config = AudioConfig {
                    sample_rate: 24000,
                    channels: 1,
                    sample_format: SampleFormat::I16,
                };
                Ok(AudioBuffer::from_pcm(decoded, config))
            }
            _ => Err(AudioError::Unsupported(format!(
                "direct decoding of {:?} not supported; use Wav or Pcm format",
                options.output_format
            ))),
        }
    }

    fn synthesize_stream(
        &self,
        text: &str,
        options: &TtsOptions,
    ) -> BoxStream<'static, AudioResult<AudioBuffer>> {
        let client = Arc::clone(&self.client);
        let text = text.to_string();
        let options = options.clone();

        let stream = async_stream::stream! {
            let tts = GoogleTts { client };
            yield tts.synthesize(&text, &options).await;
        };

        Box::pin(stream)
    }
}
